agent/tool-scripts/jmap

```diff
@@ -0,0 +1,187 @@
+#!/bin/bash
+# -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 8; sh-indentation: 8; sh-indent-for-case-alt: + -*-
+# The following script collect heap used stats for jvm and containerized jvm
+
+script_path=`dirname $0`
+script_name=`basename $0`
+pbench_bin="`cd ${script_path}/..; /bin/pwd`"
+
+# source the base script
+. "$pbench_bin"/base
+
+# Perftool scripts must provide the following functions
+# 1) Install the tool
+# 2) Start data collection
+# 3) Stop data collection
+# 4) post-process the data
+
+# Defaults
+tool=jmap
+tool_bin=/usr/bin/$tool
+group=default
+dir=""
+mode=""
+iteration="1"
+options="none"
+interval=5
+
+function usage() {
+	echo "
+	USAGE:
+	$script_name: you specified an invalid option
+	The following options are available:
+	--install,			          install this perf tool
+	--start|stop|postprocess	start/stop/post-process the data collection
+	--iteration=int 		      the iteration (required)
+	--group=str		            the perftool group (not-required)
+	--dir=str			            directory to store data collection (required)
+	--pid=int		              a process ID to profile
+	--pattern			            collect jvm stats for any PID which name matches this string
+	--interval 							    interval sleep time (5sec or higher).
+	"
+}
+
+function run_jmap() {
+	# get java pids running on host.
+	jvm_pids=$(pgrep java)
+
+	for jvm_pid in $jvm_pids; do
+		# TODO: extend the heap stats to collect min,max and non-heap
+		# check if containerized.
+		if [[ $(grep -c 'user.slice' /proc/$jvm_pid/cgroup) -eq "0" ]]; then
+			jpid=$(grep hsperfdata /proc/$jvm_pid/task/$jvm_pid/maps |awk -F "hsperfdata" '{print $2}' | sed "s/[^0-9]//g")
+			heap_used=$(nsenter --target $jvm_pid --mount --uts --ipc --net --pid $tool_bin -histo $jpid |grep Total |awk '{print $3/1024/1024}')
+		else
+			heap_used=$($tool_bin -histo $jvm_pid |grep Total |awk '{print $3/1024/1024}')
+		fi
+		# map pid to process name.
+		if [[ $(ls -A /proc/$jvm_pid/cmdline) ]]; then
+			jvm_pid=`echo "$jvm_pid-$(cat /proc/$jvm_pid/cmdline)"`
+		fi
+		echo "$(date "+%s"),$jvm_pid,$heap_used" >> "$tool_stdout_file"
+	done
+}
```

Comment:
How long does it take for jmap to attach to  a process and get the output? If it takes a long time, then this approach might have problems. What happens if you have hundreds or thousands of java processes?

As for implementation, there is a `datalog` mechanism that is used by tools that don't have a "run every so often and produce a timestamped line of output" mode. E.g. proc-vmstat's tool command is [proc-vmstat-datalog](https://github.com/distributed-system-analysis/pbench/blob/master/agent/tool-scripts/datalog/proc-vmstat-datalog). Depending on how expensive it is, you might want to keep all of the extra processing (grep/awk/mapping pids) out of the loop and deal with it in postprocessing. 

BTW, I ran jmap against the only java process I'm running on my laptop and got this:

    # time jmap -histo 2741
    2741: Unable to open socket file: target process not responding or HotSpot VM not loaded
    The -F option can be used when the target process is not responding

    real	0m5.315s
    user	0m0.221s
    sys	0m0.029s
     
and -F did not improve things. Is it designed for the HotSpot VM only? If it takes 5 seconds in the "normal" case, we'll have to rethink this approach in any case.

Reply:
It shouldn't take longer, something probably wrong with your permissions is you java process is containerised or you trying to collect it from a container ?

anyway see my results include grep and awk 
~ # time jmap -histo 1976 |grep Total |awk '{print $3/1024/1024}'
3.63048

real	0m0.277s
user	0m0.234s
sys	0m0.040s

and this is for containerised process using nester:
 ~ # time nsenter --target 6732 --mount --uts --ipc --net --pid jmap -histo 11986 |grep Total |awk '{print $3/1024/1024}'
3.63048

real	0m0.257s
user	0m0.254s
sys	0m0.043s

on the other hand, we can increase the default interval to 10 sec instead of 5. 
or we can use `sem` to protect the output file by a mutex look.

but for now looks like it's a unecessary work, and we can consider it a a follow up PR.